### PR TITLE
 config: Refactor config processing module

### LINKED ISSUE #162: target.requires breaks configuration
### Description

Version: 4.2.0
Ubuntu 20.04 LTS

### Issue request type

Configuration fails with target override for requires:
```
    "target_overrides": {
        "*": {
            "target.requires": ["ble-service-link-loss","ble-current-time-service"]
        },
```

This should work the same as `requires` in the root of the config. Currently it fails with
```
Configuring project and generating build system...
Traceback (most recent call last):
  File "/home/paul/.local/bin/mbed-tools", line 8, in <module>
    sys.exit(cli())
  File "/home/paul/.local/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/paul/.local/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/paul/.local/lib/python3.8/site-packages/mbed_tools/cli/main.py", line 38, in invoke
    super().invoke(context)
  File "/home/paul/.local/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/paul/.local/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/paul/.local/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/paul/.local/lib/python3.8/site-packages/mbed_tools/cli/build.py", line 96, in build
    generate_config(mbed_target.upper(), toolchain, program)
  File "/home/paul/.local/lib/python3.8/site-packages/mbed_tools/build/config.py", line 27, in generate_config
    config = assemble_config(target_build_attributes, program.root, program.files.app_config_file)
  File "/home/paul/.local/lib/python3.8/site-packages/mbed_tools/build/_internal/config/assemble_build_config.py", line 23, in assemble_config
    return _assemble_config_from_sources_and_lib_files(target_attributes, mbed_lib_files, mbed_app_file)
  File "/home/paul/.local/lib/python3.8/site-packages/mbed_tools/build/_internal/config/assemble_build_config.py", line 47, in _assemble_config_from_sources_and_lib_files
    return Config.from_sources(all_sources)
  File "/home/paul/.local/lib/python3.8/site-packages/mbed_tools/build/_internal/config/config.py", line 119, in from_sources
    _update_config_option(config, key, value, source)
  File "/home/paul/.local/lib/python3.8/site-packages/mbed_tools/build/_internal/config/config.py", line 133, in _update_config_option
    raise ValueError(
ValueError: Can't update option which does not exist. Attempting to set 'target.requires' to '['ble-service-link-loss', 'ble-current-time-service']' in 'File: /home/paul/gh/mbed-os-example-ble/BLE_GattServer_ExperimentalServices/mbed_app.json'.
```

- [ ] Enhancement
- [x] Bug

### PR DESCRIPTION
**Pull request by rwalton-arm**
### Description

Refactor and simplify the config processing code to make it easier to
extend and maintain.

Previously the code made assumptions about which config settings were
in different logical groups, and treated those assumptions as invariants.

We have since learned that we cannot trust those assumed invariants to
always be true, we need to refactor the code to work in a more
flexible way.

The config processing logic was also spread throughout multiple objects,
making the code hard to reason about.

Convert the `Config` object to a dictionary-like class which understands
how to perform overrides and modifications to config settings when it is
updated.

Remove the `Source` class and replace with a function that prepares the
raw config data from the JSON files in mbed-os to be passed into the
`Config` object.

The `Config` object will allow overrides to any field from targets.json,
as this is the way the old tools behaved. Our old code was constraining
the possible overrides to settings in the "config" section of the JSON
objects, along with a few hard-coded fields from targets.json we had
patched in to fix some reported issues.

Also make some performance improvements in `assemble_build_config`.
Previously we were decoding an ever increasing list of mbed_lib.json
files over and over again, then deduplicating the config settings
after parsing by relying on the uniqueness guarantees offered by
Python sets.

While I'm here, fix #162 and #155.

### Test Coverage

- [x]  This change is covered by existing or additional automated tests.
- [ ]  Manual testing has been performed (and evidence provided) as automated testing was not feasible.
- [ ]  Additional tests are not required for this change (e.g. documentation update).

### CODE REVIEW

File: src/mbed_tools/build/_internal/config/assemble_build_config.py [hunk 1]
```diff
@@ -3,88 +3,113 @@
...
-        previous_cumulative_data = current_cumulative_data
-        current_cumulative_data = CumulativeData.from_sources(all_sources)
 
-    _update_target_attributes(target_attributes, current_cumulative_data)
+def _get_file_filter_overrides(mbed_app_data: dict) -> dict:
+    return {
+        "overrides": [
+            override
+            for override in mbed_app_data.get("overrides", [])
+            if override.modifier or override.name == "requires"
```
Comment (urutva) [inline src/mbed_tools/build/_internal/config/assemble_build_config.py:78]: The `requires` is a standalone config option. Is there a scenario in which `requires` will be part of `overrides` config option?

**Review by urutva — CHANGES_REQUESTED** [review]

File: src/mbed_tools/build/_internal/config/source.py [hunk 2]
```diff
@@ -2,129 +2,153 @@
...
+        setting = ConfigSetting(
+            namespace=namespace, name=name, macro_name=macro_name, help_text=help_text, value=value,
+        )
+        # If the config item is about a certain component or feature
+        # being present, avoid adding it to the mbed_config.cmake
+        # configuration file. Instead, applications should depend on
+        # the feature or component with target_link_libraries() and the
+        # component's CMake flle (in the Mbed OS repo) will create
+        # any necessary macros or definitions.
+        if setting.name == "present":
```
[the remark this answers is not in the document]
Comment (rwalton-arm) [reply src/mbed_tools/build/_internal/config/source.py:125]: This workaround is for the `lib.present` option, which we moved in to CMake. So yeah we still need to ignore this or we'll get multiple symbol definition errors.

File: src/mbed_tools/build/_internal/config/assemble_build_config.py [hunk 3]
```diff
@@ -3,88 +3,113 @@
...
+        # We need to obtain the file filter data from mbed_app.json so we can select the correct set of mbed_lib.json
+        # files to include in the config. We don't want to update the config object with all of the app settings yet
+        # as we won't be able to apply overrides correctly until all relevant mbed_lib.json files have been parsed.
+        app_data = source.from_file(
+            mbed_app_file, default_name="app", target_filters=FileFilterData.from_config(config).labels
         )
+        _get_app_filter_labels(app_data, config)
+
+    current_filter_data = FileFilterData.from_config(config)
+    while previous_filter_data != current_filter_data:
```
[the remark this answers is not in the document]
Comment (rwalton-arm) [reply src/mbed_tools/build/_internal/config/assemble_build_config.py:48]: `mbed_lib_files` contains every `mbed_lib.json` in the mbed-os tree, then we filter the files based on the label filters in the config. We need to do multiple passes as the filter data can change after we parse a set of libs. We don't have all the filters until we've looked at all of the lib files.

File: src/mbed_tools/build/_internal/config/assemble_build_config.py (hunk 1 above)
Comment (rwalton-arm) [reply src/mbed_tools/build/_internal/config/assemble_build_config.py:78]: There's also `target.requires` which is part of the overrides.

**Review by urutva — APPROVED** [review]

**Review by Patater — APPROVED** [review]
Reviewed changes since @urutva's approval

**Pull request merged**